File: packages/sdk-python/src/wayscribe/_event.py
"""Protocol validation and final envelope fitting; delivery accounting is external."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ._config import Config
    from ._diagnostics import Diagnostics
import datetime as dt
import platform
import uuid
from collections.abc import Mapping
from dataclasses import replace

from ._capture import TOO_LARGE, UNSET, Captured, capture, json_bytes, repair_text
from ._config import valid_text
from ._errors import capture_error, public_warning
from ._version import SDK_COMMIT, SDK_NAME, __version__

OPERATIONS = frozenset(
    [
        "received",
        "identified",
        "transformed",
        "validated",
        "persisted",
        "published",
        "consumed",
        "delivered",
        "failed",
        "retried",
        "completed",
    ]
)


def now_timestamp() -> str:
    return (
        dt.datetime.now(dt.timezone.utc)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z")
    )


def new_journey_id() -> str:
    return "jrn_" + str(uuid.uuid4())


def normalize_label(value: object, diagnostics: Diagnostics) -> str | None:
    if type(value) is not str:
        diagnostics.emit("invalid_option", field="journey_label")
        return None
    value = repair_text(value)
    if not valid_text(value, max(200, len(value))):
        diagnostics.emit("invalid_option", field="journey_label")
        return None
    if len(value) > 200:
        value = value[:199] + "…"
        diagnostics.emit("invalid_option", field="journey_label", code="truncated")
    public_warning(value, "journey_label", diagnostics)
    return value


def build_envelope(
    config: Config,
    diagnostics: Diagnostics,
    *,
    journey_id: object,
    entity: object,
    operation: object,
    name: object,
    **fields: Any,
) -> bytes | None:
    try:
        return _build(config, diagnostics, journey_id, entity, operation, name, fields)
    except BaseException:
        diagnostics.emit("invalid_event", code="unreadable")
        return None


def capture_metadata(
    metadata: object, config: Config, diagnostics: Diagnostics
) -> Captured | None:
    """Snapshot validated metadata before host work, preserving capture evidence."""
    if type(metadata) is Captured:
        result = replace(metadata)
        if type(result.value) is dict:
            result.value = dict(result.value)
        return result
    try:
        if not isinstance(metadata, Mapping):
            raise ValueError()
        if len(metadata) > 1000:
            return Captured(TOO_LARGE, omitted=True)
        normalized = {}
        dropped = 0
        for index, key in enumerate(metadata):
            if index >= 1000:
                return Captured(TOO_LARGE, omitted=True)
            if type(key) is str and len(key) <= 128:
                normalized[repair_text(key)] = metadata[key]
            else:
                dropped += 1
        if dropped:
            normalized["[KEY_TOO_LONG]"] = dropped
            diagnostics.emit("invalid_option", field="metadata")
        return capture(normalized, config, field_name="metadata")
    except BaseException:
        diagnostics.emit("capture_error", field="metadata")
        return None


def _build(config, diagnostics, journey_id, entity, operation, name, fields):
    if not config.enabled:
        return None
    # Read caller-owned mappings once. Diagnostics below may reenter the host
    # and mutate them; the wire must contain exactly the strings validated here.
    entity_type = entity.get("type") if isinstance(entity, Mapping) else None
    entity_id = entity.get("id") if isinstance(entity, Mapping) else None
    if not (
        valid_text(journey_id, 128, blank=True)
        and valid_text(entity_type, 128, blank=True)
        and valid_text(entity_id, 512, blank=True)
        and (type(operation) is str)
        and (operation in OPERATIONS)
        and valid_text(name, 256, blank=True)
    ):
        diagnostics.emit("invalid_event", code="identity")
        return None
    timestamp = fields.get("timestamp", UNSET)
    if timestamp is UNSET:
        timestamp = now_timestamp()
    try:
        parsed = dt.datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            raise ValueError()
        timestamp = (
            parsed.astimezone(dt.timezone.utc)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z")
        )
    except BaseException:
        diagnostics.emit("invalid_option", field="timestamp")
        timestamp = now_timestamp()
    sdk = {"name": SDK_NAME, "version": __version__}
    if SDK_COMMIT:
        sdk["commit"] = SDK_COMMIT
    event = {
        "id": "evt_" + str(uuid.uuid4()),
        "journeyId": journey_id,
        "entity": {"type": entity_type, "id": entity_id},
        "operation": operation,
        "name": name,
        "timestamp": timestamp,
        "service": config.service,
        "environment": config.environment,
        "runtime": {
            "language": "python",
            "version": platform.python_version(),
            "sdk": sdk,
        },
    }
    if config.deployment:
        event["deployment"] = dict(config.deployment)
    for local, wire, limit in (
        ("parent_event_id", "parentEventId", 128),
        ("trace_id", "traceId", 128),
        ("span_id", "spanId", 128),
        ("message_id", "messageId", 256),
        ("correlation_id", "correlationId", 256),
    ):
        value = fields.get(local, UNSET)
        if value is UNSET:
            continue
        if (
            type(value) is str
            and len(value) <= limit
            and "\x00" not in value
            and not any(55296 <= ord(x) <= 57343 for x in value)
        ):
            event[wire] = value
        else:
            diagnostics.emit("invalid_option", field=local)
    duration = fields.get("duration_ms", UNSET)
    if type(duration) in (int, float) and duration >= 0 and (duration < float("inf")):
        event["durationMs"] = min(int(duration), 2147483647)
    elif duration is not UNSET:
        diagnostics.emit("invalid_option", field="duration_ms")
    label = fields.get("journey_label", UNSET)
    if label is not UNSET:
        label = normalize_label(label, diagnostics)
        if label is not None:
            event["journeyLabel"] = label
    error = fields.get("error", UNSET)
    if error is not UNSET and error is not None:
        event["error"] = capture_error(error, diagnostics)
    aliases = fields.get("aliases", UNSET)
    if aliases is not UNSET:
        kept = {}
        try:
            if not isinstance(aliases, Mapping) or len(aliases) > 1000:
                raise ValueError()
            for index, key in enumerate(aliases):
                if index >= 1000:
                    diagnostics.emit("invalid_option", field="aliases")
                    break
                value = aliases[key]
                if (
                    type(key) is str
                    and len(key) <= 128
                    and (type(value) is str)
                    and (len(value) <= 512)
                ):
                    kept[repair_text(key)] = repair_text(value)
                else:
                    diagnostics.emit("invalid_option", field="aliases")
        except BaseException:
            diagnostics.emit("invalid_option", field="aliases")
        event["aliases"] = kept
        display = fields.get("displayable_aliases", ())
        if type(display) not in (tuple, list):
            diagnostics.emit("invalid_option", field="displayable_aliases")
            display = ()
        usable = []
        for key in display[:1000]:
            if type(key) is str and key in kept and (key not in usable):
                usable.append(key)
                public_warning(kept[key], "displayable_alias", diagnostics)
        if usable:
            event["displayableAliases"] = usable
    captures = {}
    metadata = fields.get("metadata", UNSET)
    if metadata is not UNSET:
        result = capture_metadata(metadata, config, diagnostics)
        if result is not None:
            captures["metadata"] = result
    attempt = fields.get("attempt", UNSET)
    if type(attempt) is int and 1 <= attempt <= 9007199254740991:
        if attempt > 1:
            event["operation"] = "retried"
        if "metadata" not in captures:
            captures["metadata"] = Captured({})
        if type(captures["metadata"].value) is dict:
            captures["metadata"].value["attempt"] = attempt
        if (
            type(captures["metadata"].value) is dict
            and len(captures["metadata"].value) > 1000
        ):
            captures["metadata"] = Captured(TOO_LARGE, omitted=True)
    elif attempt is not UNSET:
        diagnostics.emit("invalid_option", field="attempt")
    if config.capture_mode != "metadata-only":
        for key in ("input", "output"):
            value = fields.get(key, UNSET)
            if value is not UNSET:
                captures[key] = (
                    replace(value)
                    if isinstance(value, Captured)
                    else capture(value, config, field_name=key)
                )
    for key, result in captures.items():
        if key != "metadata" or not result.omitted:
            event[key] = result.value

    def wire():
        return json_bytes({"protocolVersion": "0.1", "event": event})

    body = wire()
    for key in sorted(
        (key for key in ("input", "output") if key in event),
        key=lambda key: len(json_bytes(event[key])),
        reverse=True,
    ):
        if len(body) <= config.max_event_bytes:
            break
        event[key] = TOO_LARGE
        captures[key].omitted = True
        body = wire()
    if len(body) > config.max_event_bytes and "metadata" in event:
        del event["metadata"]
        captures["metadata"].omitted = True
        body = wire()
    for key, result in captures.items():
        if result.omitted:
            diagnostics.increment("payloads_omitted")
            diagnostics.emit("payload_omitted", field=key)
        elif result.truncated:
            diagnostics.increment("payloads_truncated")
            diagnostics.emit("payload_truncated", field=key)
        if result.unreadable:
            diagnostics.increment("capture_errors")
            diagnostics.emit("capture_error", field=key)
        if not result.omitted:
            for observed_name, path in result.names:
                diagnostics.report_name(
                    key,
                    observed_name,
                    f"{key}.{path}" if path else key,
                )
    if len(body) > config.max_event_bytes:
        diagnostics.emit("invalid_event", code="envelope_budget")
        return None
    return body

File: packages/sdk-go/event.go
package wayscribe

import (
	"bytes"
	"encoding/json"
	"runtime"
	"sort"
	"strings"
	"time"
	"unicode/utf8"
)

func jsonBytes(v any) []byte {
	var b bytes.Buffer
	e := json.NewEncoder(&b)
	e.SetEscapeHTML(false)
	if e.Encode(v) != nil {
		return nil
	}
	return bytes.TrimSuffix(b.Bytes(), []byte{'\n'})
}
func validOperation(o Operation) bool {
	switch o {
	case Received, Identified, Transformed, Validated, Persisted, Published, Consumed, Delivered, Failed, Retried, Completed:
		return true
	}
	return false
}
func normalizeLabel(label string, d *diagnostics) string {
	label = repairText(label)
	if strings.TrimSpace(label) == "" {
		d.emit(Diagnostic{Kind: "invalid_option", Field: "journeyLabel"})
		return ""
	}
	if utf8.RuneCountInString(label) > 200 {
		label = prefixRunes(label, 199) + "…"
		d.emit(Diagnostic{Kind: "invalid_option", Field: "journeyLabel", Code: "truncated"})
	}
	publicWarning(label, "journeyLabel", d)
	return label
}
func captureMetadata(m map[string]any, c resolvedConfig, d *diagnostics) captured {
	r := captureMetadataDeferred(m, c)
	for _, v := range r.reports {
		d.emit(v)
	}
	r.reports = nil
	return r
}
func captureMetadataDeferred(m map[string]any, c resolvedConfig) captured {
	if len(m) > 1000 {
		return captured{value: tooLarge, omitted: true}
	}
	out := map[string]any{}
	dropped := 0
	for k, v := range m {
		if utf8.RuneCountInString(k) > 128 {
			dropped++
		} else {
			out[repairText(k)] = v
		}
	}
	if dropped > 0 {
		out["[KEY_TOO_LONG]"] = dropped
	}
	r := captureValue(out, c, "metadata")
	if dropped > 0 {
		r.reports = []Diagnostic{{Kind: "invalid_option", Field: "metadata"}}
	}
	return r
}

// buildEnvelope owns capture counters only. A nil result is a refusal; the
// recorder owns admission and final recorded/sent/rejected/dropped accounting.
// Inputs must remain stable for this call. The returned bytes are SDK-owned.
func buildEnvelope(c resolvedConfig, d *diagnostics, journeyID string, entity Entity, label string, e Event) (body []byte) {
	defer func() {
		if recover() != nil {
			d.emit(Diagnostic{Kind: "invalid_event", Code: "unreadable"})
			body = nil
		}
	}()
	if !c.enabled {
		return nil
	}
	if !validText(journeyID, 128, true) || !validText(entity.Type, 128, true) || !validText(entity.ID, 512, true) || !validText(e.Name, 256, true) || !validOperation(e.Operation) {
		d.emit(Diagnostic{Kind: "invalid_event", Code: "identity"})
		return nil
	}
	id := newID("evt_")
	if id == "" {
		d.emit(Diagnostic{Kind: "invalid_event", Code: "random_source"})
		return nil
	}
	timestamp := e.Timestamp
	if timestamp.IsZero() {
		timestamp = time.Now()
	}
	if timestamp.Year() < 0 || timestamp.Year() > 9999 {
		d.emit(Diagnostic{Kind: "invalid_option", Field: "Timestamp"})
		timestamp = time.Now()
	}
	event := map[string]any{"id": id, "journeyId": journeyID, "entity": entity, "operation": e.Operation, "name": e.Name, "timestamp": timestamp.UTC().Format("2006-01-02T15:04:05.000Z"), "service": c.Service, "environment": c.Environment, "runtime": map[string]any{"language": "go", "version": runtime.Version(), "sdk": map[string]any{"name": SDKName, "version": Version}}}
	if len(c.Deployment) > 0 {
		event["deployment"] = c.Deployment
	}
	if label != "" {
		if l := normalizeLabel(label, d); l != "" {
			event["journeyLabel"] = l
		}
	}
	if e.Error != nil {
		event["error"] = captureError(*e.Error, d)
	}
	for _, v := range []struct {
		k, s string
		max  int
	}{{"parentEventId", e.ParentEventID, 128}, {"traceId", e.TraceID, 128}, {"spanId", e.SpanID, 128}, {"messageId", e.MessageID, 256}, {"correlationId", e.CorrelationID, 256}} {
		if v.s == "" {
			continue
		}
		if validText(v.s, v.max, true) {
			event[v.k] = v.s
		} else {
			d.emit(Diagnostic{Kind: "invalid_option", Field: v.k})
		}
	}
	if e.DurationMS != nil {
		if *e.DurationMS < 0 {
			d.emit(Diagnostic{Kind: "invalid_option", Field: "durationMs"})
		} else {
			event["durationMs"] = min(*e.DurationMS, maxMS)
		}
	}
	if e.Aliases != nil {
		aliases := map[string]string{}
		if e.aliasesOversize || len(e.Aliases) > 1000 {
			d.emit(Diagnostic{Kind: "invalid_option", Field: "aliases"})
		} else {
			for k, v := range e.Aliases {
				if utf8.RuneCountInString(k) <= 128 && utf8.RuneCountInString(v) <= 512 {
					aliases[repairText(k)] = repairText(v)
				} else {
					d.emit(Diagnostic{Kind: "invalid_option", Field: "aliases"})
				}
			}
		}
		event["aliases"] = aliases
		display := []string{}
		seen := map[string]bool{}
		for _, key := range e.DisplayableAliases[:min(len(e.DisplayableAliases), 1000)] {
			if v, ok := aliases[key]; ok && !seen[key] {
				display = append(display, key)
				seen[key] = true
				publicWarning(v, "displayableAliases", d)
			}
		}
		if len(display) > 0 {
			event["displayableAliases"] = display
		}
		if e.displayOversize || len(e.DisplayableAliases) > 1000 {
			d.emit(Diagnostic{Kind: "invalid_option", Field: "displayableAliases"})
		}
	}
	captures := map[string]captured{}
	if e.metadataCapture != nil {
		r := *e.metadataCapture
		for _, v := range r.reports {
			d.emit(v)
		}
		if m, ok := r.value.(map[string]any); ok {
			copy := make(map[string]any, len(m))
			for k, v := range m {
				copy[k] = v
			}
			r.value = copy
		}
		captures["metadata"] = r
	} else if e.Metadata != nil {
		captures["metadata"] = captureMetadata(e.Metadata, c, d)
	}
	if e.Attempt > 0 && int64(e.Attempt) <= safeInteger {
		if e.Attempt > 1 {
			event["operation"] = Retried
		}
		r, exists := captures["metadata"]
		if !exists {
			r = captured{value: map[string]any{}}
		}
		if m, ok := r.value.(map[string]any); ok {
			m["attempt"] = e.Attempt
			if len(m) > 1000 {
				r = captured{value: tooLarge, omitted: true}
			}
		}
		captures["metadata"] = r
	} else if e.Attempt != 0 {
		d.emit(Diagnostic{Kind: "invalid_option", Field: "attempt"})
	}
	if c.CaptureMode != MetadataOnly {
		if e.Input.present {
			captures["input"] = captureValue(e.Input.value, c, "input")
		}
		if e.Output.present {
			captures["output"] = captureValue(e.Output.value, c, "output")
		}
	}
	for k, v := range captures {
		if k != "metadata" || !v.omitted {
			event[k] = v.value
		}
	}
	wire := func() []byte { return jsonBytes(map[string]any{"protocolVersion": "0.1", "event": event}) }
	body = wire()
	keys := []string{}
	for _, k := range []string{"input", "output"} {
		if _, ok := event[k]; ok {
			keys = append(keys, k)
		}
	}
	sort.SliceStable(keys, func(i, j int) bool { return len(jsonBytes(event[keys[i]])) > len(jsonBytes(event[keys[j]])) })
	for _, k := range keys {
		if len(body) <= c.MaxEventBytes {
			break
		}
		event[k] = tooLarge
		r := captures[k]
		r.omitted = true
		captures[k] = r
		body = wire()
	}
	if len(body) > c.MaxEventBytes {
		if _, ok := event["metadata"]; ok {
			delete(event, "metadata")
			r := captures["metadata"]
			r.omitted = true
			captures["metadata"] = r
			body = wire()
		}
	}
	for _, k := range []string{"input", "output", "metadata"} {
		r, ok := captures[k]
		if !ok {
			continue
		}
		if r.omitted {
			d.increment("payloads_omitted")
			d.emit(Diagnostic{Kind: "payload_omitted", Field: k})
		} else if r.truncated {
			d.increment("payloads_truncated")
			d.emit(Diagnostic{Kind: "payload_truncated", Field: k})
		}
		if r.unreadable {
			d.increment("capture_errors")
			d.emit(Diagnostic{Kind: "capture_error", Field: k})
		}
		if !r.omitted && len(body) <= c.MaxEventBytes {
			for _, n := range r.names {
				d.emit(n)
			}
		}
	}
	if len(body) > c.MaxEventBytes {
		d.emit(Diagnostic{Kind: "invalid_event", Code: "envelope_budget"})
		return nil
	}
	return body
}
